### PR TITLE
Fix #790

### LINKED ISSUE #790: Implicit declaration of functions in configure tests
**Describe the bug**
Using the heimdal 7.5.0, 7.7.0, or even the latest code available in master today (2ee8834696f6d5c732f73005de78631005f1f527), some configure tests use functions without declaring them and/or without including the headers that declare them, resulting in incorrect determinations about system capabilities being made if the implicit declaration of functions is considered an error.

For example, some failing configure checks are:

```
checking for socket... no
checking for gethostbyname... no
```

From config.log, the reason for the failures is not that `socket` or `gethostbyname` don't exist (they do exist on macOS); you just forgot to include the right headers:

```
configure:20234: checking for socket
configure:20263: ccache /usr/bin/clang  -o conftest -pipe -Os -Werror=implicit-function-declaration -arch x86_64  -I/opt/local/include -D_LARGE_FILES= -L/opt/local/lib -Wl,-headerpad_max_install_names -arch x86_64 conftest.c      >&5
conftest.c:106:1: error: implicit declaration of function 'socket' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
socket()
^
1 error generated.
```
```
configure:20328: checking for gethostbyname
configure:20357: ccache /usr/bin/clang  -o conftest -pipe -Os -Werror=implicit-function-declaration -arch x86_64  -I/opt/local/include -D_LARGE_FILES= -L/opt/local/lib -Wl,-headerpad_max_install_names -arch x86_64 conftest.c      >&5
conftest.c:106:1: error: implicit declaration of function 'gethostbyname' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
gethostbyname()
^
1 error generated.
```

You need to `#include <sys/socket.h>` to find `socket` and `#include <netdb.h>` to find `gethostbyname`, etc.

Using `grep \ implicit` on my config.log, there are many such occurrences:

```
conftest.c:60:1: error: implicit declaration of function 'db_create' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
conftest.c:60:1: error: implicit declaration of function 'db_create' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
conftest.c:60:1: error: implicit declaration of function 'db_create' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
conftest.c:60:1: error: implicit declaration of function 'db_create' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
conftest.c:60:1: error: implicit declaration of function 'db_create' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
conftest.c:60:1: error: implicit declaration of function 'db_create' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
conftest.c:52:1: error: implicit declaration of function 'dbm_firstkey' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
conftest.c:58:8: error: implicitly declaring library function 'strchr' with type 'char *(const char *, int)' [-Werror,-Wimplicit-function-declaration]
conftest.c:106:1: error: implicit declaration of function 'socket' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
conftest.c:106:1: error: implicit declaration of function 'socket' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
conftest.c:106:1: error: implicit declaration of function 'gethostbyname' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
conftest.c:106:1: error: implicit declaration of function 'gethostbyname' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
conftest.c:106:1: error: implicit declaration of function 'syslog' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
conftest.c:106:1: error: implicit declaration of function 'syslog' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
conftest.c:108:1: error: implicit declaration of function 'gethostbyname2' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
conftest.c:108:1: error: implicit declaration of function 'gethostbyname2' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
conftest.c:108:1: error: implicit declaration of function 'gethostbyname2' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
conftest.c:159:1: error: implicit declaration of function 'bswap16' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
conftest.c:159:1: error: implicit declaration of function 'bswap32' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
conftest.c:159:1: error: implicit declaration of function 'bswap64' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
conftest.c:156:1: error: implicit declaration of function 'pidfile' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
conftest.c:156:1: error: implicit declaration of function 'pidfile' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
conftest.c:226:8: error: implicitly declaring library function 'memset' with type 'void *(void *, int, unsigned long)' [-Werror,-Wimplicit-function-declaration]
conftest.c:226:8: error: implicitly declaring library function 'memset' with type 'void *(void *, int, unsigned long)' [-Werror,-Wimplicit-function-declaration]
conftest.c:227:20: error: implicitly declaring library function 'memset' with type 'void *(void *, int, unsigned long)' [-Werror,-Wimplicit-function-declaration]
conftest.c:231:2: error: implicitly declaring library function 'memset' with type 'void *(void *, int, unsigned long)' [-Werror,-Wimplicit-function-declaration]
conftest.c:227:1: error: implicit declaration of function 'crypt' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
conftest.c:227:1: error: implicit declaration of function 'crypt' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
conftest.c:246:14: error: implicitly declaring library function 'memset' with type 'void *(void *, int, unsigned long)' [-Werror,-Wimplicit-function-declaration]
conftest.c:246:14: error: implicitly declaring library function 'memset' with type 'void *(void *, int, unsigned long)' [-Werror,-Wimplicit-function-declaration]
conftest.c:342:1: error: implicit declaration of function 'getpwnam_r' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
conftest.c:342:1: error: implicit declaration of function 'getpwnam_r' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
conftest.c:342:1: error: implicit declaration of function 'getpwuid_r' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
conftest.c:342:1: error: implicit declaration of function 'getpwuid_r' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
conftest.c:342:1: error: implicit declaration of function 'door_create' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
conftest.c:342:1: error: implicit declaration of function 'door_create' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
conftest.c:343:1: error: implicit declaration of function 'add_key' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
conftest.c:343:1: error: implicit declaration of function 'add_key' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
```

Note that not all of these functions exist on macOS, so not all of these necessarily indicate an error in heimdal.

**To Reproduce**
1. On macOS (or presumably other UNIX-like operating systems) run `CFLAGS=-Werror=implicit-function-declaration ./configure` and check the contents of config.log for occurrences of `error: implicit declaration of function`, `error: implicitly declaring library function`, or perhaps other similar errors if your compiler is not clang.

**Expected behavior**
configure tests should determine system capabilities accurately.

**Desktop (please complete the following information):**
 - OS: macOS
 - Browser n/a
 - Version n/a

**Additional context**
Implicit declaration of functions is disallowed by the C99 standard and later, though compilers have traditionally only issued a warning. That changes with the compiler included in Apple's Xcode 12 and later where that condition is now an error. If you don't have access to Xcode 12 or later, you can simulate this behavior by setting `CFLAGS=-Werror=implicit-function-declaration` when running `./configure`.

### PR DESCRIPTION
**Pull request by nicowilliams**

### CODE REVIEW

**Comment by lhoward** [comment]
Builds OK on macOS 10.15.7. I'll try with 12 when I am back on my home machine.

File: lib/sqlite/Makefile.am [hunk 1]
```diff
@@ -3,7 +3,7 @@
 include $(top_srcdir)/Makefile.am.common
 
 if ENABLE_PTHREAD_SUPPORT
-AM_CPPFLAGS += -DSQLITE_THREADSAFE=1
+AM_CPPFLAGS += -DSQLITE_THREADSAFE=1 -DSQLITE_DEFAULT_FILE_PERMISSIONS=0600
```
Comment (jaltman) [inline lib/sqlite/Makefile.am:6]: The equivalent change for Windows lib/sqlite/NTMakefile

diff --git a/lib/sqlite/NTMakefile b/lib/sqlite/NTMakefile
index f9e90fb77..f6daa119d 100644
--- a/lib/sqlite/NTMakefile
+++ b/lib/sqlite/NTMakefile
@@ -37,6 +37,10 @@ INCFILES=    \
        $(INCDIR)\sqlite3.h     \
        $(INCDIR)\sqlite3ext.h

+cdefines = $(cdefines) -DSQLITE_THREADSAFE=1 \
+       -DSQLITE_DEFAULT_FILE_PERMISSIONS=0600 \
+       -DSQLITE_WITHOUT_ZONEMALLOC=1
+
 $(LIBSQLITE): $(OBJ)\sqlite3.obj
        $(LIBCON)

**Review by jaltman — CHANGES_REQUESTED** [review]
a couple of changes required for Windows

File: lib/sqlite/Makefile.am (hunk 1 above)
Comment (nicowilliams) [reply lib/sqlite/Makefile.am:6]: Oops, I wrote:

```# pmode 0600 = _S_IREAD | _S_IWRITE

$(OBJ)\sqlite3.obj: sqlite3.c $(INCFILES)
        $(C2OBJ_P) -DSQLITE_THREADSAFE=1 -DSQLITE_DEFAULT_FILE_PERMISSIONS=0600
```

Comment (nicowilliams) [reply lib/sqlite/Makefile.am:6]: Fixed.  Please check.

**Comment by jaltman** [comment]
with the exception of tests/kdc/check-iprop all of the tests pass on macOS Apple Silicon with Apple clang version 13.0.0.

all tests pass on 32-bit Windows.

**Comment by lhoward** [comment]
check-iprop hasn't worked for a long time on macOS

**Comment by jaltman** [comment]
if the test is known to fail we should disable it on macOS so that it doesn't block the execution of the test groups that follow.

**Pull request merged**